cmd/config-encrypted.go

```diff
@@ -217,89 +249,155 @@ func migrateIAMConfigsEtcdToEncrypted(ctx context.Context, client *etcd.Client)
-		res, err := objAPI.ListObjects(GlobalContext, minioMetaBucket,
-			minioConfigPrefix, marker, "", maxObjectList)
+		res, err := objAPI.ListObjects(GlobalContext, minioMetaBucket, minioConfigPrefix, marker, "", maxObjectList)
 		if err != nil {
 			return err
 		}
 		for _, obj := range res.Objects {
-			var (
-				cdata    []byte
-				cencdata []byte
-			)
-
-			cdata, err = readConfig(GlobalContext, objAPI, obj.Name)
+			var data, plaintext []byte
+			data, err = readConfig(GlobalContext, objAPI, obj.Name)
 			if err != nil {
 				return err
 			}
 
-			var data []byte
-			// Is rotating of creds requested?
-			if activeCredOld.IsValid() {
-				data, err = decryptData(cdata, activeCredOld, globalActiveCred)
-				if err != nil {
-					if err == madmin.ErrMaliciousData {
-						return config.ErrInvalidRotatingCredentialsBackendEncrypted(nil)
-					}
-					return err
+			// The data might be:
+			//  - encrypted with the new/current master key (incomplete migration)
+			//  - encrypted with the old master key (secret key rotation)
+			//  - encrypted with the old or current secret key using the "legacy" format (migration)
+			// We first try the current master key, then the old master key and finally the legacy
+			// decryption scheme. We do this in this order, since trying a decryption with a master key
+			// is much cheaper then trying the legacy scheme decryption. (HMAC vs. Argon2id)
+			// So, the migration from "legacy" to "new" format is slightly slower - but a secret key rotation
+			// is significantly faster. Since the format migration is a one-time opeation (but key rotation is not)
+			// this seems justifable.
+			plaintext, err = key.DecryptBytes(data)
+			if err == sio.NotAuthentic {
+				if oldKey != nil {
+					plaintext, err = oldKey.DecryptBytes(data)
 				}
-			} else {
-				data = cdata
-			}
-
-			if !utf8.Valid(data) {
-				_, err = decryptData(data, globalActiveCred)
-				if err == nil {
-					// Config is already encrypted with right keys
-					continue
+				if err == sio.NotAuthentic {
+					plaintext, err = decryptData(data, oldCred, globalActiveCred) // "legacy" scheme
 				}
-				return errors.New("config data not in plain-text form or encrypted")
 			}
-
-			cencdata, err = madmin.EncryptData(globalActiveCred.String(), data)
 			if err != nil {
 				return err
 			}
 
-			if err = saveConfig(GlobalContext, objAPI, obj.Name, cencdata); err != nil {
+			data, err = key.EncryptBytes(plaintext)
+			if err != nil {
+				return err
+			}
+			if err = saveConfig(GlobalContext, objAPI, obj.Name, data); err != nil {
 				return err
 			}
 		}
 
 		if !res.IsTruncated {
 			break
 		}
-
 		marker = res.NextMarker
 	}
+	return nil
+}
 
-	if encrypted && globalActiveCred.IsValid() && activeCredOld.IsValid() {
-		logger.Info("Rotation complete, please make sure to unset MINIO_ACCESS_KEY_OLD and MINIO_SECRET_KEY_OLD envs")
+func createBackendEncrypted(ctx context.Context, objAPI ObjectLayer, backend config.Backend) error {
```

Comment:
Not really sure what this function does. I cannot tell where the encryption is taking place. Maybe doc something.

Reply:
Sure, will add a function documentation. It basically creates the `backend-encrypted` object at the backend.